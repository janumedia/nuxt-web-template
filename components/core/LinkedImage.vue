<template>
    <div :class="`${className} ${verticalAlign ? className + '--vertical-align' : ''}`">
        <nuxt-link v-if="url !== undefined && url.indexOf('http') == -1" :to="url" :title="title">
            <lazy-image 
                :src="src" 
                :title="title" 
                :square="square" 
                @onReady="imageReady" 
                @onLoaded="imageLoaded"
                @onError="imageError"
            />
            <div v-if="thumbnailMode" :class="`${className}__title`"><span>{{ title }}</span></div>
        </nuxt-link>
        <a v-else-if="url !== undefined && url.indexOf('http') > -1" :href="url" target="_blank" :title="title">
            <lazy-image 
                :src="src" 
                :title="title" 
                :square="square" 
                @onReady="imageReady" 
                @onLoaded="imageLoaded"
                @onError="imageError"
            />
        </a>
        <span v-else>
            <lazy-image 
                :src="src" 
                :title="title" 
                :square="square" 
                @onReady="imageReady" 
                @onLoaded="imageLoaded"
                @onError="imageError"
            />
        </span>
    </div>
</template>
<script>
import LazyImage from "@/components/core/LazyImage"
export default {
    props: ["src", "url", "title", "square", "thumbnailMode", "verticalAlign"],
    components: {
        LazyImage
    },
    data() {
        return {
            className: "linked-image"
        }
    },
    methods: {
        imageReady(el, index) {
            this.$emit('onReady', this, index);
        },
        imageLoaded(el, index) {
            this.$emit('onLoaded', this, index);
        },
        imageError(index) {
            this.$emit('onError', this, index);
        }
    }
}
</script>
<style lang="scss" scoped>
@import "@/assets/css/base.scss";

.linked-image {
    a {
        &:hover .linked-image__title {
            opacity: 1;
        }
    }
    &__title {
        text-align: center;
        color: rgba(255, 255, 255, 1);
        text-transform: uppercase;
        letter-spacing: 0.3em;
        font-weight: 700;
        position: absolute;
        top: 0;
        bottom: 0;
        left: 0;
        right: 0;
        text-align: center;
        width:auto;
        background: rgba(0, 0, 0, 0.1);
        @include transition(opacity 0.35s);
        span {
            position: absolute;
            top: 50%;
            left: 0;
            right: 0;
            padding: 20px;
            @include translateY(-50%);
        }
    }
    &--vertical-align {
        position:absolute;
        top:0;
        left: 0;
        right: 0;
        width:auto;
        height: 100%;
        img {
            top:0;
        }
    }
}
// mouse hover supported only. usually for non touch devices //
@media only screen and (hover:hover) {
    .linked-image {
        a {
            div {
                opacity: 0.85;
                // fix move, scale issue //
                -webkit-filter: brightness(1.01);
                -moz-filter: brightness(1.01);
                -o-filter: brightness(1.01);
                -ms-filter: brightness(1.01);
                filter: brightness(1.01);
            }
            &:hover div {
                opacity: 1;
            }
        }
        &__title {
            opacity: 0;
        }
    }
}
</style>
