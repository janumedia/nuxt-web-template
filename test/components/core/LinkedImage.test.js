import { shallowMount, mount, RouterLinkStub } from '@vue/test-utils'
import NuxtLink from '@/.nuxt/components/nuxt-link'
import LinkedImage from '@/components/core/LinkedImage'

global.process.browser = true;
jest.useFakeTimers();

const createWrapper = (propsData, mountType=shallowMount) => {
    return mountType(LinkedImage, {
        attachToDocument: true,
        propsData,
        stubs: {
            RouterLink: RouterLinkStub
        }
    })
}

describe('LinkedImage.vue', () => {
    
    let wrapper,
        IMAGE       = 'image.jpg',
        IMAGE_ERROR = 'image-error.jpg',
        propsData   = {
                        src: IMAGE,
                        title: 'image title',
                        url:'/link/to/page'
                      }

    beforeEach(() => {
        LinkedImage.components = LinkedImage.components || {};
        LinkedImage.components.NuxtLink = NuxtLink
        Object.defineProperty(Image.prototype, 'src', {
            set: function(src) {
                this._src = src;
                if(src == IMAGE_ERROR) this.onerror();
                else this.onload();
            },
            get: function() {
                return this._src;
            }
        })
        
    });

    afterEach(() => {
        //reset
        Element.prototype.getBoundingClientRect = jest.fn(() => {
            return {
                width: 0,
                height: 0,
                top: 0,
                left: 0,
                bottom: 0,
                right: 0,
            }
        });

        Object.defineProperty(Image.prototype, 'complete',  {
            value: false
        });

        //destroy and release scrollListener
        wrapper.destroy();

        //reset setTimer
        jest.clearAllTimers();
    })
    
    it('should render LinkedImage without url', () => {
        wrapper = createWrapper({src:propsData.src, title:propsData.title}, mount);
        expect(wrapper.vm.$el).toMatchSnapshot();
    });
    
    it('should render LinkedImage with router-link', () => {
        wrapper = createWrapper(propsData);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render LinkedImage with http url', () => {
        wrapper = createWrapper({...propsData, url:'http://domain.name'});
        expect(wrapper.html()).toMatchSnapshot();
    });
    
    it('should render LinkedImage with square mode', () => {
        wrapper = createWrapper({...propsData, square:true}, mount);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render LinkedImage with thumbnail mode', () => {
        wrapper = createWrapper({...propsData, thumbnailMode:true}, mount);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render LinkedImage with verticalAlign', () => {
        wrapper = createWrapper({...propsData, verticalAlign:true}, mount);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should emit onReady and onLoaded event', () => {
        Object.defineProperty(Image.prototype, 'complete',  {
            value: true
        })
        wrapper = createWrapper({...propsData}, mount);
        jest.runOnlyPendingTimers();
        expect(wrapper.emitted().onReady).toBeTruthy();
        expect(wrapper.emitted().onLoaded).toBeTruthy();
        expect(wrapper.emitted().onError).toBeFalsy();
    });

    it('should emit onReady and onError event', () => {
        Object.defineProperty(Image.prototype, 'complete',  {
            value: true
        })
        wrapper = createWrapper({...propsData, src:IMAGE_ERROR}, mount);
        jest.runOnlyPendingTimers();
        expect(wrapper.emitted().onReady).toBeTruthy();
        expect(wrapper.emitted().onLoaded).toBeFalsy();
        expect(wrapper.emitted().onError).toBeTruthy();
    });
})