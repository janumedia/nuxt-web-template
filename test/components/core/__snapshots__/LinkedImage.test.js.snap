// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedImage.vue should render LinkedImage with http url 1`] = `
<div class="linked-image ">
  <a href="http://domain.name" target="_blank" title="image title">
    <lazyimage-stub src="image.jpg" title="image title"></lazyimage-stub>
  </a>
</div>
`;

exports[`LinkedImage.vue should render LinkedImage with router-link 1`] = `
<div class="linked-image ">
  <nuxtlink-stub to="/link/to/page" title="image title">
    <lazyimage-stub src="image.jpg" title="image title"></lazyimage-stub>
    <!---->
  </nuxtlink-stub>
</div>
`;

exports[`LinkedImage.vue should render LinkedImage with square mode 1`] = `
<div class="linked-image ">
  <a title="image title">
    <div class="lazy-image">
      <div>
        <img src="/images/assets/2x2-placeholder.svg" data-src="image.jpg" alt="image title" data-pin-nopin="true">
      </div>
    </div>
    <!---->
  </a>
</div>
`;

exports[`LinkedImage.vue should render LinkedImage with thumbnail mode 1`] = `
<div class="linked-image ">
  <a title="image title">
    <div class="lazy-image">
      <div>
        <img src="/images/assets/3x2-placeholder.svg" data-src="image.jpg" alt="image title" data-pin-nopin="true">
      </div>
    </div>
    <div class="linked-image__title"><span>image title</span></div>
  </a>
</div>
`;

exports[`LinkedImage.vue should render LinkedImage with verticalAlign 1`] = `
<div class="linked-image linked-image--vertical-align">
  <a title="image title">
    <div class="lazy-image">
      <div>
        <img src="/images/assets/3x2-placeholder.svg" data-src="image.jpg" alt="image title" data-pin-nopin="true">
      </div>
    </div>
    <!---->
  </a>
</div>
`;

exports[`LinkedImage.vue should render LinkedImage without url 1`] = `
<div
  class="linked-image "
>
  <span>
    <div
      class="lazy-image"
    >
      <div>
        <img
          alt="image title"
          data-pin-nopin="true"
          data-src="image.jpg"
          src="/images/assets/3x2-placeholder.svg"
        />
      </div>
    </div>
  </span>
</div>
`;
